### PR TITLE
Add domain for the product micro-service

### LINKED ISSUE #28: Make product domain
**Description**
Make product domain

**Why is this needed**
A product domain is needed

**Tasks**
- [x] Think about options for the domain (can be expanded later)
- [ ] Implement domain
- [ ] Test domain

### PR DESCRIPTION
**Pull request by Rubenvdbrink**
## Changes
Added domain and tests for Product Micro-Service

## Checklist
- [x] Make domain
- [x] Make tests
- [x] add jacoco to pom

## When merging, this pull request 
closes #28

### CODE REVIEW

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        if (weight < 0) throw new IllegalArgumentException("Product weight has to be bigger than 0");
+        this.weight = weight;
+    }
+
+    /**
+     * Sets the nutritional value {@link NutritionalValue} of the product.
+     * @param nutritionalValue - nutritional values of the product.
+     */
+    public void setNutritionalValue(NutritionalValue nutritionalValue) {
+        Objects.requireNonNull(nutritionalValue, "The product needs to have nutritional values");
```
Comment (JustMilan) [inline product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java:54]: Maybe check for Empty too?
requireNonBlankOrEmpty(nutririonalValue);

**Review by JustMilan — CHANGES_REQUESTED** [review]

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java (hunk 1 above)
Comment (Rubenvdbrink) [reply product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java:54]: Nutritional Value can't be emtpy, this is already checked in the Nutritional Value constructor

File: product/src/main/java/nl/softwarestrijders/waiter/product/common/Generated.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * When this annotation is used the function below will be ignored by jacoco
+ */
+@Documented
+@Retention(RUNTIME)
+@Target({TYPE, METHOD})
+public @interface Generated {
+}
```
Comment (xandervedder) [inline product/src/main/java/nl/softwarestrijders/waiter/product/common/Generated.java:18]: F.Y.I. the reason this works;
> Classes and methods annotated with annotation whose retention policy is runtime or class and whose simple name is Generated are filtered out during generation of report - JaCoCo

In my opinion it pollutes the codebase a little bit, but alrighty

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/NutritionalValue.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Record that contains all nutritional values of the {@link Product}.
+ */
+public record NutritionalValue(int kcal, int fats, int carbs, int proteins, int salts) {
+    /**
+     * Constructor that checks if any value is smaller than 0. if true, throw exception.
+     */
+    public NutritionalValue {
+        for(var value : List.of(kcal, fats, carbs, proteins, salts))
+            if(value < 0) throw new IllegalArgumentException("Nutritional value has to be bigger than 0.");
```
Comment (xandervedder) [inline product/src/main/java/nl/softwarestrijders/waiter/product/domain/NutritionalValue.java:14]: nit:
```suggestion
        for (var value : List.of(kcal, fats, carbs, proteins, salts))
            if (value < 0) throw new IllegalArgumentException("Nutritional value has to be bigger than 0.");
```

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    @Generated
+    public double getPrice() {return price;}
+    @Generated
+    public String getDescription() {return description;}
+    @Generated
+    public String getName() {return name;}
+    @Generated
+    public NutritionalValue getNutritionalValue() {return nutritionalValue;}
+    @Generated
+    public int getWeight() {return weight;}
```
Comment (xandervedder) [inline product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java:98]: Personally (this is why linting is kinda nice.. I blame myself) I would format with intellij, and see what comes out of it.

File: product/src/test/java/nl/softwarestrijders/waiter/product/domain/NutritionalValueTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    private static Stream<Arguments> wrongNutritionalValues() {
+        return Stream.of(
+                Arguments.of(-1,0,0,0,0),
+                Arguments.of(0,-1,0,0,0),
+                Arguments.of(0,0,-1,0,0),
+                Arguments.of(0,0,0,-1,0),
+                Arguments.of(0,0,0,0,-1),
+                Arguments.of(-1,-1,-1,-1,-1)
```
Comment (xandervedder) [inline product/src/test/java/nl/softwarestrijders/waiter/product/domain/NutritionalValueTest.java:31]: nit:
```suggestion
                Arguments.of(-1, 0, 0, 0, 0),
                Arguments.of(0, -1, 0, 0, 0),
                Arguments.of(0, 0, -1, 0, 0),
                Arguments.of(0, 0, 0, -1, 0),
                Arguments.of(0, 0, 0, 0, -1),
                Arguments.of(-1, -1, -1, -1, -1)
```

File: product/src/test/java/nl/softwarestrijders/waiter/product/domain/NutritionalValueTest.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        );
+    }
+
+    private static Stream<Arguments> goodNutritionalValues() {
+        return Stream.of(
+                Arguments.of(0,0,0,0,0),
+                Arguments.of(1,1,1,1,1),
+                Arguments.of(100,100,100,100,100)
```
Comment (xandervedder) [inline product/src/test/java/nl/softwarestrijders/waiter/product/domain/NutritionalValueTest.java:39]: nit: same as above

**Review by xandervedder — CHANGES_REQUESTED** [review]

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java (hunk 4 above)
Comment (Rubenvdbrink) [reply product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java]: `    @Generated
    public UUID getId() {
        return id;
    }

    @Generated
    public double getPrice() {
        return price;
    }

    @Generated
    public String getDescription() {
        return description;
    }

    @Generated
    public String getName() {
        return name;
    }

    @Generated
    public NutritionalValue getNutritionalValue() {
        return nutritionalValue;
    }

    @Generated
    public int getWeight() {
        return weight;
    }`

**Review by JustMilan — APPROVED** [review]

File: product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java (hunk 4 above)
Comment (xandervedder) [reply product/src/main/java/nl/softwarestrijders/waiter/product/domain/Product.java]: Yes those are collapsed, are they not? if you expand them they would look like this, right:
```java
    @Generated
    public String getName() {
        return name;
    }
```

**Review by xandervedder — APPROVED** [review]

**Pull request merged**